packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -538,162 +539,118 @@ impl<'m> RunnerImpl<'m> {
         })
     }
 
-    unsafe fn new_buffer(&self, ptr: *const u8, len: usize, _capacity: usize) -> Buffer {
-        let s = std::slice::from_raw_parts(ptr, len);
-        s.into()
-    }
-
     /// TODO: DOC, flushing from a single column
     fn array_data_from_js(
         &mut self,
         mv8: &'m MiniV8,
         data: &mv8::Value<'m>,
-        dt: &DataType,
+        field: &Field,
         len: Option<usize>,
     ) -> Result<ArrayData> {
         // `data` must not be dropped until flush is over, because
         // pointers returned from FFI point inside `data`'s ArrayBuffers' memory.
-        let obj = data
-            .as_object()
-            .ok_or_else(|| Error::Embedded("Flush data not object".into()))?;
-        let child_data: mv8::Array<'_> = obj.get("child_data")?;
+        let obj = data.as_object().ok_or_else(|| {
+            Error::Embedded(format!("Flush data not object for field {:?}", field))
+        })?;
 
         // `data_node_from_js` isn't recursive -- doesn't convert children.
         let data: mv8::DataFfi = mv8.data_node_from_js(data);
 
-        let n_children = child_data.len();
-        let child_data: Vec<ArrayData> = match dt.clone() {
-            DataType::List(field) => {
-                let child: mv8::Value<'_> = child_data.get(0)?;
-                Ok(vec![self.array_data_from_js(
+        // This target length is used because the JS repr does not mirror buffer building as Rust
+        // Arrow and pyarrow.
+        let target_len = len.unwrap_or(data.len);
+
+        // TODO: Extra copies when generating buffers, because JS Arrow doesn't align things
```

Comment:
Is this TODO still valid? I don't quite understand it but you have seemed to change it since the original code

Reply:
Yes, we are not using the original pointer from JavaScript, thus we copying the data. Previously this was done by `.into()` (one reason I don't like the trait, it's hard to see clones), now it's done by `Buffer::from_slice_ref`.

Reply:
Can we rephrase the TODO then. So the TODO isn't extra copies, it's to _stop making_ extra copies?

Reply:
Also the reasoning you've said isn't explained in the TODO so it would be good to do so.